types/vision/index.d.ts

```diff
@@ -1,212 +1,271 @@
+        runtimeOptions?: RuntimeOptions;
+        /** the content type of the engine results. Defaults to 'text/html'. */
+        contentType?: string;
+        /** specify whether the engine compile() method is 'sync' or 'async'. Defaults to 'sync'. */
+        compileMode?: 'sync' | 'async';
+        /**
+         * A global context used with all templates.
+         * The global context option can be either an object or a function that takes the request as its only argument and returns a context object.
+         * The request object is only provided when using the view handler or reply.view().
+         * When using server.render() or request.render(), the request argument will be null.
+         * When rendering views, the global context will be merged with any context object specified on the handler or using reply.view().
+         * When multiple context objects are used, values from the global context always have lowest precedence.
+         */
+        context?: object | ((request: Request) => object);
     }
-}
 
-/**
- * Options for initialising server views manager
- * @see {@link https://github.com/hapijs/vision/blob/master/API.md#serverviewsoptions}
- */
-export interface ServerViewsConfiguration extends ServerViewsAdditionalOptions {
-    /** required object where each key is a file extension (e.g. 'html', 'hbr'), mapped to the npm module used for rendering the templates. Alternatively, the extension can be mapped to an object */
-    engines: {[fileExtension: string]: NpmModule} | ServerViewsEnginesOptions;
-    /** defines the default filename extension to append to template names when multiple engines are configured and no explicit extension is provided for a given template. No default value. */
-    defaultExtension?: string;
-}
+    /**
+     * Options passed to module when compiling template.
+     * Cast your options to this interface or extend it with:
+     *      declare module 'hapi' {
+     *          interface CompileOptions {
+     *              noEscape: boolean;
+     *          }
+     *      }
+     */
+    type CompileOptions = object;
+    type RuntimeOptions = object;
+
+    /**
+     * The rendering function. The required function signature depends on the compileMode settings (see below).
+     *
+     * If compileMode is 'sync', the signature is compile(template, options),
+     * the return value is a function with signature function(context, options) (the compiled sync template),
+     * and the method is allowed to throw errors.
+     *
+     * If compileMode is 'async',
+     * the signature is compile(template, options, next) where next has the signature function(err, compiled),
+     * compiled is a function with signature function(context, options, callback) (the compiled async template)
+     * and callback has the signature function(err, rendered).
+     */
+    interface ServerViewCompile {
+        (template: string, options: any): (context: any, options: any) => void;
```

Comment:
Again, this may cause issues with TS not narrowing types correctly, you want to create a union type of functions for this.

Reply:
Like this?

```ts
type ServerViewCompile = (template: string, options: any, next?: ServerViewCompileNext) => (context: any, options: any) => void | void;
```

Reply:
Not exactly:
```
((template: string, options: any) => (context: any, options: any) => void) | (template: string, options: any, next: ServerViewCompileNext) => void);
```

Reply:
Ah, yeah. Missed that.